File: parking_system/engines/allocation_engine.py
from typing import Dict, List, Optional
from datetime import datetime
import uuid

from ..domain.zone import Zone
from ..domain.parking_request import ParkingRequest, ParkingRequestState, ParkingRequestError
from ..domain.parking_slot import ParkingSlot, ParkingSlotError


class AllocationError(Exception):
    pass


class OperationRecord:
    def __init__(
        self,
        operation_type: str,
        request_id: str,
        slot_id: str,
        prev_slot_state: bool,
        prev_request_state: ParkingRequestState,
    ):
        self.operation_id: str = str(uuid.uuid4())
        self.operation_type: str = operation_type
        self.request_id: str = request_id
        self.slot_id: str = slot_id
        self.prev_slot_state: bool = prev_slot_state
        self.prev_request_state: ParkingRequestState = prev_request_state
        self.timestamp: datetime = datetime.utcnow()


class AllocationEngine:
    def __init__(self, zones: Dict[str, Zone]):
        self._zones: Dict[str, Zone] = zones
        self._operations: List[OperationRecord] = []

    # ---------- Allocation ----------
    def allocate(self, request: ParkingRequest) -> None:
        if request.state != ParkingRequestState.VALIDATED:
            raise AllocationError("Request must be VALIDATED to allocate")

        # Step 1: try same-zone allocation
        zone = self._zones.get(request.preferred_zone_id)
        slot = self._find_available_slot(zone) if zone else None

        # Step 2: fallback to other zones
        if not slot:
            for z in self._zones.values():
                if z.zone_id == request.preferred_zone_id:
                    continue
                slot = self._find_available_slot(z)
                if slot:
                    break

        if not slot:
            request.transition_to(ParkingRequestState.FAILED)
            raise AllocationError("No slots available in any zone")

        # Step 3: record previous state
        op = OperationRecord(
            operation_type="ALLOCATE",
            request_id=request.request_id,
            slot_id=slot.slot_id,
            prev_slot_state=slot.is_available,
            prev_request_state=request.state,
        )
        self._operations.append(op)

        # Step 4: perform allocation
        slot.allocate(request.vehicle_id)
        request.transition_to(ParkingRequestState.ALLOCATED)
        request._allocated_zone_id = slot.area_id  # temporary direct binding
        request._allocated_slot_id = slot.slot_id

    def _find_available_slot(self, zone: Zone) -> Optional[ParkingSlot]:
        for area in zone.areas:
            for slot in area.available_slots:
                return slot
        return None

    # ---------- Release ----------
    def release(self, request: ParkingRequest) -> None:
        if request.state not in {ParkingRequestState.ALLOCATED, ParkingRequestState.ACTIVE}:
            raise AllocationError("Request must be ALLOCATED or ACTIVE to release")

        zone = self._zones.get(request.allocated_zone_id)
        if not zone:
            raise AllocationError("Zone not found for allocated request")

        area = zone.get_area(request.allocated_zone_id)
        slot = area.get_slot(request.allocated_slot_id)

        # record previous state
        op = OperationRecord(
            operation_type="RELEASE",
            request_id=request.request_id,
            slot_id=slot.slot_id,
            prev_slot_state=slot.is_available,
            prev_request_state=request.state,
        )
        self._operations.append(op)

        # perform release
        slot.release()
        request.transition_to(ParkingRequestState.COMPLETED)
    def allocate(self, request: ParkingRequest) -> None:
            if request.state != ParkingRequestState.VALIDATED:
                raise AllocationError("Request must be VALIDATED to allocate")

            # Step 1: preferred zone first
            preferred_zone = self._zones.get(request.preferred_zone_id)
            slot = self._find_available_slot(preferred_zone) if preferred_zone else None

            # Step 2: cross-zone fallback with penalty scoring
            if not slot:
                sorted_zones = sorted(
                    (z for z in self._zones.values() if z.zone_id != request.preferred_zone_id),
                    key=lambda z: getattr(z, "penalty", 1),  # default penalty=1
                )
                for z in sorted_zones:
                    slot = self._find_available_slot(z)
                    if slot:
                        break

            # Step 3: no slots anywhere → FAILED
            if not slot:
                request.transition_to(ParkingRequestState.FAILED)
                raise AllocationError("No slots available in any zone")

            # Step 4: record operation
            op = OperationRecord(
                operation_type="ALLOCATE",
                request_id=request.request_id,
                slot_id=slot.slot_id,
                prev_slot_state=slot.is_available,
                prev_request_state=request.state,
            )
            self._operations.append(op)

            # Step 5: perform allocation
            slot.allocate(request.vehicle_id)
            request.transition_to(ParkingRequestState.ALLOCATED)
            request._allocated_zone_id = slot.area_id
            request._allocated_slot_id = slot.slot_id

File: parking_system/src/lib/parking/AllocationEngine.ts
/**
 * AllocationEngine Class
 * 
 * Core allocation logic with same-zone priority and cross-zone fallback.
 * 
 * Allocation Algorithm:
 * 1. Try to find available slot in preferred zone (same-zone priority)
 * 2. If not found, search adjacent zones first (lower penalty)
 * 3. If still not found, search remaining zones (higher penalty)
 * 4. Return allocation result with slot and penalty information
 * 
 * Time Complexity:
 *   - allocate: O(z * a * s) where z = zones, a = areas, s = slots
 * Space Complexity: O(1) additional
 */

import { Zone } from './Zone';
import { ParkingSlot } from './ParkingSlot';
import { ParkingRequest } from './ParkingRequest';
import { Vehicle } from './Vehicle';
import { 
  IAllocationOperation, 
  RequestStatus, 
  SlotStatus,
  generateId 
} from './types';

export interface AllocationResult {
  success: boolean;
  slot: ParkingSlot | null;
  zone: Zone | null;
  penalty: number;
  message: string;
  operation: IAllocationOperation | null;
}

export class AllocationEngine {
  private zones: Map<string, Zone>;
  private basePenalty: number;

  constructor(zones: Zone[], basePenalty: number = 10) {
    this.zones = new Map(zones.map(zone => [zone.id, zone]));
    this.basePenalty = basePenalty;
  }

  /**
   * Update zones reference
   */
  public updateZones(zones: Zone[]): void {
    this.zones = new Map(zones.map(zone => [zone.id, zone]));
  }

  /**
   * Main allocation method
   * Implements same-zone priority with cross-zone fallback
   */
  public allocate(
    request: ParkingRequest,
    vehicle: Vehicle
  ): AllocationResult {
    // Validate request is in REQUESTED state
    if (request.status !== RequestStatus.REQUESTED) {
      return {
        success: false,
        slot: null,
        zone: null,
        penalty: 0,
        message: `Cannot allocate: request is in ${request.status} state`,
        operation: null,
      };
    }

    const preferredZone = this.zones.get(request.preferredZoneId);
    if (!preferredZone) {
      return {
        success: false,
        slot: null,
        zone: null,
        penalty: 0,
        message: `Preferred zone ${request.preferredZoneId} not found`,
        operation: null,
      };
    }

    // Step 1: Try preferred zone first (same-zone priority)
    let slot = preferredZone.findAvailableSlot(vehicle.type);
    if (slot) {
      return this.completeAllocation(request, slot, preferredZone, 0);
    }

    // Step 2: Try adjacent zones (lower penalty)
    for (const adjZoneId of preferredZone.adjacentZoneIds) {
      const adjZone = this.zones.get(adjZoneId);
      if (adjZone) {
        slot = adjZone.findAvailableSlot(vehicle.type);
        if (slot) {
          const penalty = this.basePenalty * adjZone.penaltyMultiplier;
          return this.completeAllocation(request, slot, adjZone, penalty);
        }
      }
    }

    // Step 3: Try all other zones (highest penalty)
    for (const [zoneId, zone] of this.zones) {
      if (zoneId === request.preferredZoneId) continue;
      if (preferredZone.adjacentZoneIds.includes(zoneId)) continue;

      slot = zone.findAvailableSlot(vehicle.type);
      if (slot) {
        const penalty = this.basePenalty * zone.penaltyMultiplier * 2;
        return this.completeAllocation(request, slot, zone, penalty);
      }
    }

    // No slot available anywhere
    return {
      success: false,
      slot: null,
      zone: null,
      penalty: 0,
      message: 'No available slots in any zone',
      operation: null,
    };
  }

  /**
   * Complete the allocation by updating slot and request states
   */
  private completeAllocation(
    request: ParkingRequest,
    slot: ParkingSlot,
    zone: Zone,
    penalty: number
  ): AllocationResult {
    // Save previous states for rollback
    const prevSlotStatus = slot.status;
    const prevRequestStatus = request.status;

    // Update slot state
    if (!slot.allocate(request.id)) {
      return {
        success: false,
        slot: null,
        zone: null,
        penalty: 0,
        message: 'Failed to allocate slot',
        operation: null,
      };
    }

    // Update request state
    if (!request.allocate(slot.id, zone.id, penalty)) {
      // Rollback slot on request failure
      slot.restoreState(prevSlotStatus, null);
      return {
        success: false,
        slot: null,
        zone: null,
        penalty: 0,
        message: 'Failed to update request state',
        operation: null,
      };
    }

    // Create operation record
    const operation: IAllocationOperation = {
      id: generateId('op'),
      requestId: request.id,
      slotId: slot.id,
      previousSlotStatus: prevSlotStatus,
      previousRequestStatus: prevRequestStatus,
      operationType: 'ALLOCATE',
      timestamp: new Date(),
    };

    const isCrossZone = request.isCrossZone();
    const message = isCrossZone
      ? `Allocated to ${zone.name} (cross-zone, penalty: ${penalty})`
      : `Allocated to ${zone.name}`;

    return {
      success: true,
      slot,
      zone,
      penalty,
      message,
      operation,
    };
  }

  /**
   * Mark a request as occupied (vehicle arrived)
   */
  public occupy(request: ParkingRequest): IAllocationOperation | null {
    if (request.status !== RequestStatus.ALLOCATED || !request.allocatedSlotId) {
      return null;
    }

    const slot = this.findSlotById(request.allocatedSlotId);
    if (!slot) return null;

    const prevSlotStatus = slot.status;
    const prevRequestStatus = request.status;

    if (!slot.occupy() || !request.occupy()) {
      return null;
    }

    return {
      id: generateId('op'),
      requestId: request.id,
      slotId: slot.id,
      previousSlotStatus: prevSlotStatus,
      previousRequestStatus: prevRequestStatus,
      operationType: 'OCCUPY',
      timestamp: new Date(),
    };
  }

  /**
   * Release a parking slot (vehicle departed)
   */
  public release(request: ParkingRequest): IAllocationOperation | null {
    if (request.status !== RequestStatus.OCCUPIED || !request.allocatedSlotId) {
      return null;
    }

    const slot = this.findSlotById(request.allocatedSlotId);
    if (!slot) return null;

    const prevSlotStatus = slot.status;
    const prevRequestStatus = request.status;

    if (!slot.release() || !request.release()) {
      return null;
    }

    return {
      id: generateId('op'),
      requestId: request.id,
      slotId: slot.id,
      previousSlotStatus: prevSlotStatus,
      previousRequestStatus: prevRequestStatus,
      operationType: 'RELEASE',
      timestamp: new Date(),
    };
  }

  /**
   * Cancel a request (from REQUESTED or ALLOCATED state)
   */
  public cancel(request: ParkingRequest): IAllocationOperation | null {
    if (!request.canTransitionTo(RequestStatus.CANCELLED)) {
      return null;
    }

    const slotId = request.allocatedSlotId;
    let slot: ParkingSlot | undefined;
    let prevSlotStatus = SlotStatus.AVAILABLE;

    if (slotId) {
      slot = this.findSlotById(slotId);
      if (slot) {
        prevSlotStatus = slot.status;
        slot.cancelAllocation();
      }
    }

    const prevRequestStatus = request.status;
    request.cancel();

    return {
      id: generateId('op'),
      requestId: request.id,
      slotId: slotId ?? '',
      previousSlotStatus: prevSlotStatus,
      previousRequestStatus: prevRequestStatus,
      operationType: 'CANCEL',
      timestamp: new Date(),
    };
  }

  /**
   * Find slot by ID across all zones
   */
  private findSlotById(slotId: string): ParkingSlot | undefined {
    for (const zone of this.zones.values()) {
      const slot = zone.getSlotById(slotId);
      if (slot) return slot;
    }
    return undefined;
  }

  /**
   * Get zone by ID
   */
  public getZone(zoneId: string): Zone | undefined {
    return this.zones.get(zoneId);
  }

  /**
   * Get all zones
   */
  public getAllZones(): Zone[] {
    return Array.from(this.zones.values());
  }
}
